lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -535,19 +535,52 @@ class WhenOpVisitor : public LastConnectResolver<WhenOpVisitor> {
   /// scope, i.e. not in a WhenOp region, then there is no condition.
   Value andWithCondition(Operation *op, Value value) {
     // 'and' the value with the current condition.
-    return OpBuilder(op).createOrFold<AndPrimOp>(
+    OpBuilder builder(op);
+    return andWithCondition(builder, value);
+  }
+
+  /// And a 1-bit value with the current condition.  If we are in the outer
+  /// scope, i.e. not in a WhenOp region, then there is no condition.
+  Value andWithCondition(OpBuilder &builder, Value value) {
+    // 'and' the value with the current condition.
+    return builder.createOrFold<AndPrimOp>(
         condition.getLoc(), condition.getType(), condition, value);
   }
 
-  // Create an implication from the condition to the given value
-  // This is implemented as (or (not condition) value) to avoid
-  // the corner case where value is an implication.
-  Value impliesLTLCondition(Operation *op, Value value) {
-    OpBuilder builder = OpBuilder(op);
-    Value notCond = builder.createOrFold<LTLNotIntrinsicOp>(
-        condition.getLoc(), condition.getType(), condition);
-    return builder.createOrFold<LTLOrIntrinsicOp>(
-        condition.getLoc(), condition.getType(), notCond, value);
+  // Or a 1-bit value with the current condition.  If we are in the outer
+  /// scope, i.e. not in a WhenOp region, then there is no condition.
+  Value orWithCondition(OpBuilder &builder, Value value) {
+    // 'or' the value with the current condition.
+    return builder.createOrFold<OrPrimOp>(
+        condition.getLoc(), condition.getType(), condition, value);
+  }
+
+  template <typename Op>
+  void visitVerifIntrinsic(Op op) {
+    OpBuilder builder(op);
+    Value foldedWhen;
+    LTLClockIntrinsicOp clockOp;
+    LTLDisableIntrinsicOp disable;
+    // Extract the clock and disable from the UInt<1>
+    if ((clockOp =
+             dyn_cast<LTLClockIntrinsicOp>(op.getProperty().getDefiningOp()))) {
+      if ((disable = dyn_cast<LTLDisableIntrinsicOp>(
+               clockOp.getInput().getDefiningOp()))) {
+
```

Comment:
What happens if the nesting is flipped, e.g. `disable(clock(P))`? The condition should probably be folded into the disable here as well, shouldn't it?

Reply:
Also, what happens if this is just `assert(P)` without a disable? Should this create a disable like `assert(disable(P))` with the when condition?

Reply:
For the case where assert has no disable, I think that it's totally legal to add a disable if it's wrapped in a when

Reply:
I added a canonical ordering for clock and disable and renamed disable's operands so that they make sense and are not just `lhs` and `rhs`, but instead `input` and `disable`